R/select_nse.R

```diff
@@ -0,0 +1,509 @@
+    "names" = .select_names(x, data, ignore_case, regex, verbose),
+    "starts_with" = ,
+    "ends_with" = ,
+    "matches" = ,
+    "contains" = ,
+    "regex" = .select_helper(x, data, ignore_case, regex, verbose),
+    .select_context(x, data, ignore_case, regex, verbose)
+  )
+}
+
+# e.g 1:3, or gear:cyl
+.select_seq <- function(expr, data, ignore_case, regex, verbose) {
+  x <- .eval_expr(
+    expr[[2]],
+    data = data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+  y <- .eval_expr(
+    expr[[3]],
+    data = data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+  x:y
+}
+
+# e.g -cyl
+.select_minus <- function(expr, data, ignore_case, regex, verbose) {
+  x <- .eval_expr(
+    expr[[2]],
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+  if (length(x) == 0L) {
+    seq_along(data)
+  } else {
+    x * -1L
+  }
+}
+
+# e.g c("gear", "cyl")
+.select_c <- function(expr, data, ignore_case, regex, verbose) {
+  lst_expr <- as.list(expr)
+  lst_expr[[1]] <- NULL
+  unlist(lapply(
+    lst_expr,
+    .eval_expr,
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  ))
+}
+
+# e.g -(gear:cyl)
+.select_bracket <- function(expr, data, ignore_case, regex, verbose) {
+  .eval_expr(
+    expr[[2]],
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+}
+
+# e.g myvector[3]
+.select_square_bracket <- function(expr, data, ignore_case, regex, verbose) {
+  first_obj <- .eval_expr(
+    expr[[2]],
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+  .eval_expr(
+    first_obj[expr[[3]]],
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = verbose
+  )
+}
+
+.select_names <- function(expr, data, ignore_case, regex, verbose) {
+  first_obj <- .dynEval(expr[[2]], inherits = FALSE, minframe = 0L)
+  .eval_expr(
+    names(first_obj),
+    data,
+    ignore_case = ignore_case,
+    regex = regex,
+    verbose = FALSE
+  )
+}
+
+# e.g starts_with("Sep")
+.select_helper <- function(expr, data, ignore_case, regex, verbose) {
+  lst_expr <- as.list(expr)
+
+  # need this if condition to distinguish between starts_with("Sep") (that we
+  # can use directly) and starts_with(i) (where we need to get i)
+  if (length(lst_expr) == 2L && typeof(lst_expr[[2]]) == "symbol") {
+    collapsed_patterns <- .dynGet(lst_expr[[2]], inherits = FALSE, minframe = 0L)
+  } else {
+    collapsed_patterns <- paste(unlist(lst_expr[2:length(lst_expr)]),
+      collapse = "|"
+    )
+  }
+
+  rgx <- if (lst_expr[[1]] == "starts_with") {
+    paste0("^(", collapsed_patterns, ")")
+  } else if (lst_expr[[1]] == "ends_with") {
+    paste0("(", collapsed_patterns, ")$")
+  } else if (lst_expr[[1]] == "contains") {
+    paste0("(", collapsed_patterns, ")")
+  } else if (lst_expr[[1]] == "regex") {
```

Comment:
Instead of repeatedly subsetting with `[[`, you can just assign `lst_expr[[1]]` to a variable and use that instead.